lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,239 @@
+    } else if (formatChar == secondCode) {
+      return dt.second;
+    }
+    throw AssertionError("Unreachable, the key is checked in the constructor");
+  }
+
+  /// Parse a string [formattedDateTime] to a local [DateTime] as specified in the
+  /// [pattern], substituting missing values with a default. Throws an exception
+  /// on failure to parse.
+  DateTime decode(String formattedDateTime, {bool isUtc = false}) {
+    return _decode(formattedDateTime, isUtc, true);
+  }
+
+  /// Same as [decode], but will not throw on parsing erros, instead using
+  /// the default value as if the format char was not present in the [pattern].
+  DateTime tryDecode(String formattedDateTime, {bool isUtc = false}) {
+    return _decode(formattedDateTime, isUtc, false);
+  }
+
+  DateTime _decode(
+    String formattedDateTime,
+    bool isUtc,
+    bool throwOnError,
+  ) {
+    var characters = formattedDateTime.codeUnits;
+    var year = 0;
+    var century = 0;
+    var decade = 0;
+    var month = 1;
+    var day = 1;
+    var hour = 0;
+    var minute = 0;
+    var second = 0;
+    for (int i = 0; i < _blocks.length; i++) {
+      var char = _blocks.formatCharacters[i];
+      var num = _extractNumFromString(characters, i, throwOnError);
+      if (num != null) {
+        if (char == yearCode) {
+          year = num;
+        } else if (char == centuryCode) {
+          century = num;
+        } else if (char == decadeCode) {
+          decade = num;
+        } else if (char == monthCode) {
+          month = num;
+        } else if (char == dayCode) {
+          day = num;
+        } else if (char == hourCode) {
+          hour = num;
+        } else if (char == minuteCode) {
+          minute = num;
+        } else if (char == secondCode) {
+          second = num;
+        }
+      }
+    }
+    var totalYear = year + 100 * century + 10 * decade;
+    if (isUtc) {
+      return DateTime.utc(totalYear, month, day, hour, minute, second, 0, 0);
+    } else {
+      return DateTime(totalYear, month, day, hour, minute, second, 0, 0);
+    }
+  }
+
+  int? _extractNumFromString(
+      List<int> characters, int index, bool throwOnError) {
+    var parsed =
+        tryParse(characters, _blocks.starts[index], _blocks.ends[index]);
+    if (parsed == null && throwOnError) {
+      throw FormatException(
+          '${String.fromCharCodes(characters)} should only contain digits');
+    }
+    return parsed;
+  }
+
+  static final zeroCode = '0'.codeUnitAt(0);
+  int? tryParse(List<int> characters, int start, int end) {
+    int result = 0;
+    for (var i = start; i < end; i++) {
+      var character = characters[i];
+      if (character >= zeroCode && character < zeroCode + 10) {
+        result = result * 10 + (character - zeroCode);
+      } else {
+        return null;
+      }
+    }
+    return result;
+  }
+}
+
+class _ParsedFormatBlocks {
+  final formatCharacters = <int>[];
+  final starts = <int>[];
+  final ends = <int>[];
```

Comment:
To save a little space, I'd use a single list here, and do lookups by multiplying by 3.
```dart
  final _records = <int>[];
  int length = 0;
  void saveBlock(int?char, int start, int end) {
     if (char != null) {
       _records..add(char)..add(start)..add(end);
       length++;
     }
   }
   int char(int i) => _records[i * 3];
   int start(int i) => _records[i * 3 + 1];
   int end(int i) => _records[i * 3 + 2];
```
(I care, maybe too much, about avoiding allocations.)

Reply:
Keep what we have now. I can always play around with optimizations later. 

(I prefer to minimize allocations, and three growable lists have three times the potential over-allocation as one list, *but* this is for the format object, which is probably allocated just once, and then used several times. And if it turns out someone keeps allocating the same format string, we can start caching.)